### ISSUE TITLE
[AnimatePresence] Issues with spring animations

### ISSUE THREAD

**Issue by nandorojo** [nandorojo/moti#41]
When using exit animations + springs + `AnimatePresence`, there are bugs where the exited item doesn't always exit. This only happens when you have more than one style key set.

I think it might be due to this line: https://github.com/nandorojo/moti/blob/f6060e4a812dfbe092156b9865c5caf03872a32e/packages/core/src/use-map-animate-to-style.ts#L292

Example:

```tsx
<AnimatePresence>
  <MotiView animate={{ opacity: 1, translateY: 0 }} animate={{ opacity: 0, translateY: -1 }} />
</AnimatePresence>
```

The current logic is, once the last Reanimated animation callback has fired, we call `safeToUnmount` from framer motion. The problem is, if we're using springs, different animations will not take the same amount of time to animate necessarily.

This is the code now:

```ts
if (isExiting) {
  //   // if this is true, then we've finished our exit animations
  const isLastStyleKeyToAnimate = index + 1 === Object.keys(mergedStyles || {}).length
  if (isLastStyleKeyToAnimate) {
    runOnJS(reanimatedSafeToUnmount)()
  }
}
```

## Possible solution

In the case above, `translateY` could finish before `opacity`, for instance. Rather than simple call `safeToUnmount` after the last animation in the dictionary finishes, we should probably be keeping memory of every animation that completes. After the last one completes, then we know we're ready to unmount.

```ts
// outside of for-loop
const needsExiting = Object.keys(exitStyles).reduce((acc, key) => ({ ...acc, [key]: true }), {})

// ...

if (isExiting) {
  needsExiting[key] = false
  const isLastStyleKeyToAnimate = !Object.values(needsExiting).some(Boolean)
  if (isLastStyleKeyToAnimate) {
    runOnJS(reanimatedSafeToUnmount)()
  }
}
```

It would maybe make more sense to use `Set` here, but not sure if that's supported by Reanimated, so this should be fine.

This might just be related to using `2.0.0-rc.0`, which has issues with springs. I'll have to see if using stable reanimated v2 solves it.

**Comment by godness84** [comment]
Good catch! Moreover I don't understand why reanimated v2 does not call a callback when every animation (in the same `useAnimatedStyle`) ends. When they will support it, I think that it will be possible to implement something similar to `framer-motion` with `afterChildren` and `beforeChildren` flags to better orchestrate animations in children.

**Comment by nandorojo** [comment]
>  I don't understand why reanimated v2 does not call a callback when every animation (in the same useAnimatedStyle) ends.

It's a bit tough, because animations are cancellable, so given that they can all be tracking different shared values, there's never really a single instance of all animations "finishing". They can all have their own timing and lifecycle. 

The only use-case of them all finishing in tandem I can think of is exit animations. For exits, you're certain this is the last time each key will animate.

I'm not familiar with how the children props work, but I can look into it.

**Comment by nandorojo** [comment]
This should be solved in `0.7.2`. 

`yarn add moti@next`

If anyone could test this out before I merge to latest, would be useful.
